Rewrite of MCLC

PadowYT2 opened the pull request:
### This Pull Request has a lot of changes
- Usage of
  - Typescript (which translates into ESM and CJS during building)
  - Vitest
  - ESLint
  - Prettier
  - Yarn (instead of npm)
- Many functions were completly remade to work with TS and just function
- Removed Auth functions

### There are still things to finish, such as
- [x] Config
- [x] Tests
- [x] README
- [x] GitHub workflows
  - [x] Vitest
  - [x] Prettier

PadowYT2 committed:
Testing conditions

PadowYT2 committed:
Multiple initial changes

- Use yarn
- Reconfigure eslint
- Install prettier and format the code

PadowYT2 committed:
Add progress to testing

PadowYT2 committed:
components -> src

PadowYT2 committed:
[AUTH] Rewrite request into axios

PadowYT2 committed:
Finish rewriting request into axios

PadowYT2 committed:
Remove the need for checksum package

PadowYT2 committed:
Fully remove checksum

PadowYT2 committed:
Support ESM (WIP, need to convert into TS to support both)

PadowYT2 committed:
Preparations for rewriting to TS

PadowYT2 committed:
[WIP] Recode into TS

PadowYT2 committed:
'config' is never reassigned

PadowYT2 committed:
Fix logging

PadowYT2 committed:
Fixed some things about the memory

PadowYT2 committed:
Update handler.ts

PadowYT2 committed:
Unbundle launch to install and start

PadowYT2 committed:
Fix having MC the MCLC prefix

PadowYT2 committed:
Fix MC not properly downloading natives on Windows

PadowYT2 committed:
Bring back java check

PadowYT2 committed:
Uncomment this also

PadowYT2 committed:
[WIP] Forge installer

PadowYT2 committed:
[WIP] Forge instlaler

PadowYT2 committed:
Remove forge mod loader

PadowYT2 committed:
Finish all mod loaders (except Neoforge)

PadowYT2 committed:
Some outside of code changes

PadowYT2 committed:
Remove cheerio

PadowYT2 committed:
Remove Java check on installation

PadowYT2 committed:
Vanilla testing

PadowYT2 committed:
Whoops

PadowYT2 committed:
Legacy and modern versions testing

PadowYT2 committed:
Change imports of config

PadowYT2 committed:
Remove tests from tsconfig

PadowYT2 committed:
[WIP] Fix types (not working)

PadowYT2 committed:
Save for later to rework on

PadowYT2 committed:
Finish the rewrite

PadowYT2 committed:
Merge remote-tracking branch 'upstream/master'

PadowYT2 committed:
v3.19.0

PadowYT2 committed:
Formatting

Bricklou commented on .eslintrc.js:
```diff
@@ -0,0 +1,27 @@
+    parserOptions: {
+        ecmaVersion: 'latest',
+        ecmaFeatures: {
+            jsx: true,
```
Why are you enabled JSX ?

Bricklou commented on .eslintrc.js:
```diff
@@ -0,0 +1,27 @@
+    rules: {
+        eqeqeq: 'error',
+        'prettier/prettier': ['error', {}, { usePrettierrc: true }],
+        '@typescript-eslint/no-explicit-any': 0,
```
If you need this, that means you forgot to add types somewhere.

Bricklou commented on README.md:
```diff
@@ -1,220 +1,172 @@
-![logo](https://owo.whats-th.is/8mT5kxc.png)
-##### Project rewrite coming soon™
-[![Build Status](https://travis-ci.com/Pierce01/MinecraftLauncher-core.svg?branch=master)](https://travis-ci.com/Pierce01/MinecraftLauncher-core)
```
If you want to keep the build badge, you can probably use the ones from GitHub action (if you configured something)

Bricklou commented on README.md:
```diff
@@ -1,220 +1,172 @@
-    authorization: Authenticator.getAuth("username", "password"),
-    root: "./minecraft",
+const launcher = new Client();
+const options = {
```
What about a `defineConfig` (like the one you can see in vitejs) to ensure what's passed in the options ?

Bricklou commented on README.md:
```diff
@@ -1,220 +1,172 @@
+| ------------------------------ | ------- | -------------------------------------------------------------------------------------------------------------------------------- | -------- |
+| `options.removePackage`        | Boolean | Option to remove the client package zip file after its finished extracting.                                                      | False    |
+| `options.root`                 | String  | Path where you want the launcher to work in. `C:/Users/user/AppData/Roaming/.mc`                                                 | True     |
+| `options.cache`                | String  | Path where launcher files will be cached in. `C:/Users/user/AppData/Roaming/.mc/cache`                                           | False    |
```
Why don't you use the official Minecraft resources (if they are present) to cache resources?

Bricklou commented on jest.config.ts:
```diff
@@ -0,0 +1,8 @@
+import type { JestConfigWithTsJest } from 'ts-jest';
+
+const jestConfig: JestConfigWithTsJest = {
```
Jest is really not made to work with typescript and is a pain to use with it. I would recommend you to use something else like Vitest which handles typescript natively.

Bricklou commented on package.json:
```diff
@@ -1,42 +1,61 @@
+    ],
+    "type": "module",
+    "scripts": {
+        "clean": "rm -rf ./lib/",
```
Doesn't work on Windows

Bricklou commented on package.json:
```diff
@@ -1,42 +1,61 @@
+    "type": "module",
+    "scripts": {
+        "clean": "rm -rf ./lib/",
+        "build": "yarn clean && yarn build:esm && yarn tsc",
```
Why yarn ? The original project was working on NPM only

PadowYT2 replied:
npm isn't that good, while Yarn is better. pnpm might be even better, but totally not npm

Bricklou replied:
Let Pierce decide on this part

Bricklou commented on src/handler.ts:
```diff
@@ -0,0 +1,685 @@
+import { Agent as https } from 'node:https';
+import { join, resolve, sep } from 'node:path';
+import Zip from 'adm-zip';
+import axios from 'axios';
```
Axios is nice, but is something as big is required?

Bricklou commented on src/handler.ts:
```diff
@@ -0,0 +1,685 @@
+        return parsedVersion;
+    }
+
+    const manifest = `${config.url?.meta}/mc/game/version_manifest.json`;
```
Be careful, you might end with an `undefined` at the beginning of the string if not handled properly.

Bricklou commented on src/handler.ts:
```diff
@@ -0,0 +1,685 @@
+const getJar = async () => {
+    await downloadAsync(
+        parsedVersion.downloads.client.url,
+        config.directory!,
```
Never use `!`, this assumes you are certain that the type is never undefined (never assume anything). This means either your checks aren't right or your object isn't types properly.

Bricklou commented on src/handler.ts:
```diff
@@ -0,0 +1,685 @@
+    log('debug', 'Downloaded assets');
+};
+
+const parseRule = (lib: libType) => {
```
`libType` should be in PascalCase

Bricklou commented on src/handler.ts:
```diff
@@ -0,0 +1,685 @@
+import {
```
You should really split the code in multiple separated files. It is too long.

Bricklou commented on src/handler.ts:
```diff
@@ -0,0 +1,685 @@
+                '.jar',
+                `-${jarEnding}.jar`,
+            );
+            json.libraries[0].downloads.artifact.url = `https://files.minecraftforge.net/maven/${json.libraries[0].downloads.artifact.path}`;
```
You might want to extract the URL into a constant.

Bricklou commented on src/handler.ts:
```diff
@@ -0,0 +1,685 @@
+    return result;
+};
+
+const getLaunchOptions = async (modification: customLibType | null): Promise<string[]> => {
```
`customLibType` should be in PascalCase

Bricklou commented on src/handler.ts:
```diff
@@ -0,0 +1,685 @@
+    if (typeof config.memory.min === 'number' && typeof config.memory.max === 'number') {
+        if (config.memory.max < config.memory.min) {
+            log('debug', 'MIN memory is higher then MAX! Resetting!');
+            setConfig('memory', { min: Math.pow(2, 9), max: Math.pow(2, 10) });
```
I still don't get why you're doing a global variable, this is not a clean way to keep the state.

Bricklou commented on src/index.ts:
```diff
@@ -0,0 +1,4 @@
+export * as Client from './launcher';
+export { default as offline } from './authenticator';
```
If you need to rename the default, then why not just use `export` instead if `export default`

Bricklou commented on src/launcher.ts:
```diff
@@ -0,0 +1,184 @@
+                jvm.push(`-Dlog4j.configurationFile=${resolve(join(configPath, 'log4j2_112-116.xml'))}`);
+            } else if (intVersion >= 7) {
+                await downloadAsync(
+                    'https://launcher.mojang.com/v1/objects/dd2b723346a8dcd48e7f4d245f6bf09e98db9696/log4j2_17-111.xml',
```
Move into a constant

Bricklou commented on src/utils/index.ts:
```diff
@@ -0,0 +1,49 @@
+const popString = (path: string) => path.split('/').slice(0, -1).join('/');
+const cleanUp = (array: string[]) => [...new Set(Object.values(array).filter((value) => value !== null))];
+
+const getOS = () => {
```
The return type should be strongly typed

Bricklou commented on src/utils/types.ts:
```diff
@@ -0,0 +1,300 @@
+        /**
+         * Max amount of memory being used by Minecraft.
+         */
+        min: string | number;
```
\`${number}M\` instead of string ?

Bricklou commented on src/utils/types.ts:
```diff
@@ -0,0 +1,300 @@
+    client_token: string;
+    uuid: string;
+    name: string;
+    user_properties: Partial<any> | string;
```
Maybe put something else instead of `any`

Bricklou commented on src/utils/types.ts:
```diff
@@ -0,0 +1,300 @@
+    features: Record<string, boolean>;
+};
+
+export type artifactType = {
```
PascalCase

Bricklou commented on src/utils/types.ts:
```diff
@@ -0,0 +1,300 @@
+    url: string;
+};
+
+export type libType = {
```
PascalCase

Bricklou commented on src/utils/types.ts:
```diff
@@ -0,0 +1,300 @@
+    }[];
+};
+
+export type customArtifactType = {
```
PascalCase

Bricklou commented on src/utils/types.ts:
```diff
@@ -0,0 +1,300 @@
+    size?: number;
+};
+
+export type customLibType = {
```
PascalCase

Bricklou reviewed (changes requested):
The whole code looks nice, a refactor would be nice to clean the overall code.

PadowYT2 committed:
Remove JSX from ESLint (accidentally added)

PadowYT2 committed:
Use .cjs instead of .js for ESLint

https://eslint.org/docs/latest/use/configure/configuration-files#configuration-file-formats

PadowYT2 committed:
[WIP] Rework on the config system

PadowYT2 committed:
[WIP] Use vitest instead of Jest

PadowYT2 committed:
Clean build folder via an another way

PadowYT2 committed:
PascalCase

PadowYT2 committed:
No usage of ! and use Infinity by default for maxSockets

PadowYT2 committed:
Remake the exporting of `offline`

PadowYT2 committed:
Use `${number}M/G` as a type for memory

PadowYT2 committed:
Enforce OS type

PadowYT2 committed:
Remove user_properties

PadowYT2 committed:
Unbundle types

PadowYT2 committed:
Formatting

PadowYT2 committed:
Usage of custom paths

PadowYT2 committed:
Formatting

PadowYT2 committed:
Remove useless counter

PadowYT2 committed:
Undo: Remove useless counter

PadowYT2 committed:
Use a class for the Counter

PadowYT2 committed:
Remove the need to cache the version

As it is unused

PadowYT2 committed:
[WIP] FileDownloader

PadowYT2 committed:
Fix retry and return a promise

PadowYT2 committed:
Move first forge library into a constant

PadowYT2 committed:
Fix importing the library

PadowYT2 committed:
[WIP] Fix (sort of) the FileDownloader

PadowYT2 committed:
Remove FileDownloader

PadowYT2 committed:
Remake the custom paths

PadowYT2 committed:
Use rimraf

PadowYT2 committed:
Use prebuild instead of clean

PadowYT2 committed:
Refactor on README.md

PadowYT2 committed:
Make use of classes again. Finish configuration

PadowYT2 committed:
Adapt README to the latest commit

PadowYT2 committed:
[WIP] Create a workflow

PadowYT2 committed:
Formatting

PadowYT2 committed:
Bump workflow's NodeJS version to v18

PadowYT2 committed:
Bump NodeJS from v16 to v18

PadowYT2 committed:
Move workflow runner into the repository's folder

PadowYT2 committed:
Fix workflows

PadowYT2 committed:
Update tests.yml

PadowYT2 committed:
Fully fix workflows (looks cluttered, need help)

PadowYT2 committed:
Setup Java

PadowYT2 committed:
Testing

PadowYT2 committed:
Fix workflows for the last time

PadowYT2 committed:
Add workflow badge to README

PadowYT2 committed:
Rewrite handler to a class (+fixing config)

PadowYT2 committed:
Testing

PadowYT2 committed:
Update tests.yml

PadowYT2 committed:
More testing

PadowYT2 committed:
Launch MC manually

PadowYT2 committed:
-Dfml.earlyprogresswindow=false

PadowYT2 committed:
Testing -Dorg.lwjgl.glfw.libname=glfw_wayland

PadowYT2 committed:
Finish with workflows

PadowYT2 committed:
Remove unused imports

PadowYT2 committed:
Proper LICENSE update

my bad

PadowYT2 committed:
Remove deleted file from .eslintignore

PadowYT2 committed:
Remove browser env in ESLint config

PadowYT2 committed:
Comment in .gitignore

PadowYT2 committed:
Remove JSX from Prettier config

PadowYT2 committed:
Remove unused type and add default for maxSockets

PadowYT2 committed:
Array<string> -> string[]

PadowYT2 committed:
forgot about this 1

PadowYT2 commented:
Looking back at this, I realized that making a launcher in NodeJS isn't a great idea. Tauri would be a better solution here and there are libraries that already exist (not trying to advertise)! Even if you want to keep using NodeJS, there's new tech by now and most of the packages used can be improved.

I'll be closing this PR as I'm not looking forward into making a launcher as of right now and most likely never in NodeJS. The code itself is kinda a mess if you try to actually read/edit it. In theory, the fork should still work, but some updates should be done from upstream.

:wave:
